### PR TITLE
Avoiding stats for unwanted scenarios

### PR DESCRIPTION
**Pull request by sivashanmukh**
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/363)

### CODE REVIEW

File: jobsdb/jobsdb.go [hunk 1]
```diff
@@ -1361,9 +1361,6 @@ func (jd *HandleT) markClearEmptyResult(ds dataSetT, stateFilters []string, cust
 }
 
 func (jd *HandleT) isEmptyResult(ds dataSetT, stateFilters []string, customValFilters []string, parameterFilters []ParameterFilterT) bool {
-	queryStat := stats.NewJobsDBStat("isEmptyCheck", stats.TimerType, jd.tablePrefix)
-	queryStat.Start()
-	defer queryStat.End()
```
Comment (sivashanmukh) [inline jobsdb/jobsdb.go:1366]: This was added by me when I didn't realize it was just an in-memory operation. These stats are not necessary.

File: jobsdb/jobsdb.go [hunk 2]
```diff
@@ -1415,6 +1412,13 @@ parameterFilters do a AND query on values included in the map
 */
 func (jd *HandleT) getProcessedJobsDS(ds dataSetT, getAll bool, stateFilters []string,
 	customValFilters []string, limitCount int, parameterFilters []ParameterFilterT) ([]*JobT, error) {
+	jd.checkValidJobState(stateFilters)
+
+	if jd.isEmptyResult(ds, stateFilters, customValFilters, parameterFilters) {
+		logger.Debugf("[getProcessedJobsDS] Empty cache hit for ds: %v, stateFilters: %v, customValFilters: %v, parameterFilters: %v", ds, stateFilters, customValFilters, parameterFilters)
+		return []*JobT{}, nil
+	}
+
```
Comment (sivashanmukh) [inline jobsdb/jobsdb.go:1421]: We were collecting query-stats even when we don't run a query. This is a fix for that.

File: jobsdb/jobsdb.go [hunk 3]
```diff
@@ -1537,6 +1534,11 @@ parameterFilters do a AND query on values included in the map
 */
 func (jd *HandleT) getUnprocessedJobsDS(ds dataSetT, customValFilters []string,
 	order bool, count int, parameterFilters []ParameterFilterT) ([]*JobT, error) {
+	if jd.isEmptyResult(ds, []string{NotProcessed.State}, customValFilters, parameterFilters) {
+		logger.Debugf("[getUnprocessedJobsDS] Empty cache hit for ds: %v, stateFilters: NP, customValFilters: %v, parameterFilters: %v", ds, customValFilters, parameterFilters)
+		return []*JobT{}, nil
+	}
+
```
Comment (sivashanmukh) [inline jobsdb/jobsdb.go:1541]: We were collecting query-stats even when we don't run a query. This is a fix for that.

File: jobsdb/jobsdb.go [hunk 4]
```diff
@@ -2535,6 +2529,10 @@ func (jd *HandleT) GetUnprocessed(customValFilters []string, count int, paramete
 	queryStat = stats.NewJobsDBStat(statName+"unprocessed", stats.TimerType, jd.tablePrefix)
 	queryStat.Start()
 	defer queryStat.End()
+
+	//The order of lock is very important. The mainCheckLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
```
Comment (sivashanmukh) [inline jobsdb/jobsdb.go:2535]: Moving the comment back to its rightful place.

File: jobsdb/jobsdb.go [hunk 5]
```diff
@@ -2584,6 +2579,10 @@ func (jd *HandleT) GetProcessed(stateFilter []string, customValFilters []string,
 	queryStat = stats.NewJobsDBStat(statName+"processed", stats.TimerType, jd.tablePrefix)
 	queryStat.Start()
 	defer queryStat.End()
+
+	//The order of lock is very important. The mainCheckLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
```
Comment (sivashanmukh) [inline jobsdb/jobsdb.go:2585]: Moving the comment back to its rightful place.

File: processor/processor.go [hunk 6]
```diff
@@ -903,7 +903,7 @@ func (proc *HandleT) mainLoop() {
 	currLoopSleep := time.Duration(0)
 
 	for {
-		if proc.handlePendingGatewayJobs() {
+		if !proc.handlePendingGatewayJobs() {
```
Comment (sivashanmukh) [inline processor/processor.go:906]: This is the main reason this PR was raised. Needs a careful review.

Comment (snarkychef) [reply processor/processor.go:906]: Looks good. 

My only comment is I don't prefer negative if conditions. Makes it hard to read.

Comment (sivashanmukh) [reply processor/processor.go:906]: Instead of the whole logic below, how about using https://github.com/cenkalti/backoff/

**Pull request merged**